CSharpSourceCode/CampaignSupport/TORWanderersCampaignBehavior.cs

```diff
@@ -0,0 +1,90 @@
+﻿using HarmonyLib;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using TaleWorlds.CampaignSystem;
+using TaleWorlds.CampaignSystem.Actions;
+using TaleWorlds.CampaignSystem.SandBox.CampaignBehaviors;
+using TaleWorlds.Core;
+using TaleWorlds.ObjectSystem;
+using TOW_Core.Utilities.Extensions;
+
+namespace TOW_Core.CampaignSupport
+{
+    public class TORWanderersCampaignBehavior : CampaignBehaviorBase
+    {
+        public override void SyncData(IDataStore dataStore)
+        {
+        }
+
+        public override void RegisterEvents()
+        {
+            CampaignEvents.AfterSettlementEntered.AddNonSerializedListener(this, OnAfterSettlementEntered);
+        }
+
+        private void OnAfterSettlementEntered(MobileParty mobileParty, Settlement settlement, Hero hero)
+        {
+            if (hero == null || !hero.IsHumanPlayerCharacter)
+            {
+                return;
+            }
+
+            //check for unsuitable wanderers
+            for (int i = 0; i < settlement.HeroesWithoutParty.Count; i++)
+            {
+                var wanderer = settlement.HeroesWithoutParty[i];
+                if (wanderer != null && wanderer.Occupation == Occupation.Wanderer && wanderer.Culture != settlement.Culture)
+                {
+                    //look for empty suitable settlement to move unsuitable wanderer
+                    var suitableTown = (from x in Town.AllTowns
+                                        where x.Settlement.Culture == wanderer.Culture
+                                        orderby x.Settlement.HeroesWithoutParty.Count ascending
+                                        select x).FirstOrDefault().Settlement;
+                    if (suitableTown != null)
+                    {
+                        EnterSettlementAction.ApplyForCharacterOnly(wanderer, settlement);
+                    }
+                    else
+                    {
+                        LeaveSettlementAction.ApplyForCharacterOnly(wanderer);
+                        wanderer.ChangeState(Hero.CharacterStates.NotSpawned);
+                    }
+
+                    //create suitable wanderer
+                    CharacterObject template = settlement.Culture.NotableAndWandererTemplates.Where(h => h.Occupation == Occupation.Wanderer).GetRandomElementInefficiently();
+                    Hero newWanderer = HeroCreator.CreateSpecialHero(template, settlement, null, null, Campaign.Current.Models.AgeModel.HeroComesOfAge + 5 + MBRandom.RandomInt(27));
+                    AdjustEquipmentImp(newWanderer.BattleEquipment);
+                    AdjustEquipmentImp(newWanderer.CivilianEquipment);
+                    newWanderer.ChangeState(Hero.CharacterStates.Active);
+                    EnterSettlementAction.ApplyForCharacterOnly(newWanderer, settlement);
```

Comment:
I think we should only create a new suitable wanderer if there isn't any matching one in the settlement at all (none of HeroesWithoutParties are matching). We are going to spam too many wanderers like this. Maybe move this piece of code outside of the for loop and only do it once instead for every unsuitable wanderer?